### DISCUSSION TITLE
How to print the used environment variable names in runtime?

### DISCUSSION THREAD

**Discussion by petroslamb** [pydantic/pydantic#5724] [Question]
Hi, lots of work here, good job keeping up.

I have a rather complicated case of `BaseSettings` that has sub models other `BaseSettings` classes. Each of them has a different `env_prefix` and I am trying to understand the environment variables that are loaded at runtime. Is there a way to get them from the instantiated object? (v1.10.7)

Thanks!

**Comment by chrisK824** [comment]
Hey @petroslamb , not sure if I get the question, if you could provide a piece of code that would be great.

If it is just the print of attributes of an instance you could use `__dict__` method on the instance object or even `__dict__.keys()`

Update:

What about to just print the cast `.dict()` of Settings class or instance, just like in the docs https://docs.pydantic.dev/latest/usage/settings/

**Comment by petroslamb** [comment]
Hi @chrisK824 ,

Sorry for not being clear. 

I am looking for a way given a `BaseSettings` instance, to get the env vars that it picks up. For example if an imaginary method `_construct_env_vars` existed, then calling it would return them. 

```
from pydantic import BaseSettings

class MyConfig(BaseSettings):
    v1: str
    
    class Config:
        env_prefix = 'TEST_'
```

Then when using:
```
config = MyConfig()
print(config._construct_env_vars()) 
# or perhaps better 
print(MyConfig._construct_env_vars()  
```
Would return something that helps me inspect them:
```
{'v1': `TEST_V1`}
```
, or something similar, that would allow me to easily infer them.

I needed it while trying debug the following [issue](https://github.com/pydantic/pydantic/issues/5728) in a similar complex config and was not sure how to inspect them.

**Reply by chrisK824** [reply]
May I ask how are you creating those in the first place? If those are created on the fly with some method, then this should be the place you could add some serialiser method perhaps to verbosely print them as well.

**Reply by petroslamb** [reply]
Thanks for the answer.

So, the env vars pre-exist in the environment, I am not creating them in any way. 
I am looking for a way to inspect the BaseSettings derived class (or instance) and see which ones it tries to load in the python runtime.

**Comment by chrisK824** [comment]
I'm still not sure I understand, but I updated my answer with the recommended way to print out Settings from pydantic @petroslamb

**Comment by petroslamb** [comment]
@chrisK824 from the example you updated, the `.dict()` method prints:

```
print(Settings().dict())
```
```
{
    'auth_key': 'xxx',
    'api_key': 'xxx',
    'redis_dsn': RedisDsn('redis://user:pass@localhost:6379/1', ),
    'pg_dsn': PostgresDsn('postgres://user:pass@localhost:5432/foobar', ),
    'amqp_dsn': AmqpDsn('amqp://user:pass@localhost:5672/', scheme='amqp',
user='user', password='pass', host='localhost', host_type='int_domain',
port='5672', path='/'),
    'special_function': <built-in function cos>,
    'domains': set(),
    'more_settings': {'foo': 'bar', 'apple': 1},
}
```
By default, the environment variable name is built by concatenating the prefix and field name.

For example, to override special_function above, you could use:
```
  export my_prefix_special_function='foo.bar'
```

I am looking for a way to inpect the `Settings()` object and find this environment variable final name `my_prefix_special_function` that corresponds to the field name `special_function`. I need to see the constructed one (`my_prefix_` + `special_function`) and not just the first or last parts seperately. This will help me debug a complex version of this `Settings` class that I have, where the construction of the env var name is not straightforward.

**Comment by slobberchops** [comment]
I think this question is trying to address a similar problem that I am trying to solve. It is not easy to determine the complete set of environment variables (and their names) that are available for use for a given program.

In my case, someone wrote a set of unit tests that are driven by `BaseSettings` classes and those may contain nested references to `BaseModel` classes. It is very difficult to enumerate the set of available environment variables that are available for use. For example, I might want to write a simple script that given a set of `BaseSettings` classes it will show me all environment variable names available for use.

So if I have defined:

```python

class MyServer(pydantic.BaseModel):
  host: str
  port: int

class MySettings(pydantic.BaseSettings):
  class Config:
    env_prefix = 'TEST_'
    env_nested_delimiter = '__'
  file_location: str = 'default-file-location'
  server1: MyServer
  server2: MyServer
```

I might learn the available environment variables by running a script:


```sh
% show-env-vars mymodule.MySettings
TEST_FILE_LOCATION
TEST_SERVER1__HOST
TEST_SERVER1__PORT
TEST_SERVER2__HOST
TEST_SERVER2__PORT
```

The short of it I guess is, "is there a way to enumerate the environment variable names accepted by settings class X"?

@petroslamb Is this a similar problem to what you are trying to solve?

**Reply by petroslamb** [reply]
Hi @slobberchops,

Yes " enumerate the environment variable names accepted by settings class" this is the problem indeed. I do not mind having a script, I was thinking to inspect the class from python. So I was asking if there is a way to do so, perhaps by accessing some property or calling a method on the class.

**Comment by inozemtsev** [comment]
Hi @petroslamb 

Looks like `EnvSettingsSource` [class](https://github.com/pydantic/pydantic-settings/blob/main/pydantic_settings/sources.py#L359) handles loading env variables.

There is this `_extract_field_info` [method](https://github.com/pydantic/pydantic-settings/blob/main/pydantic_settings/sources.py#L141C9-L141C9) which returns environment variable name of the field as one of the results.

So, something like this should do the trick, I think:

```python
from pydantic_settings import BaseSettings, EnvSettingsSource

class Example(BaseSettings):
    class Config:
        env_prefix: str = 'example_'

    env_variable: str = 'var'

example = Example()

env_settings = EnvSettingsSource(
            example.__class__,
            case_sensitive=example.model_config['case_sensitive'],
            env_prefix=example.model_config['env_prefix'],
            env_nested_delimiter=example.model_config['env_nested_delimiter']
        )

for name, field_info in example.model_fields.items():
    env_var_name = env_settings._extract_field_info(field_info, name)[0][1]
    print(name, env_var_name)
    
# env_variable example_env_variable
```

**Reply by niderhoff** [reply]
That is great @inozemtsev, but for nested models it unfortunately needs some recursion added.

**Reply by niderhoff** [reply]
I modified it so that it is recursive and takes the delimiter into account:

```py
def extract_env_vars(model: BaseModel, submodel_name: Optional[str] = None):
    def extract_env_vars_inner(model: BaseModel, submodel_name: Optional[str] = None):
        env_vars = []
        for name, field_info in model.model_fields.items():
            if field_info.default_factory:
                env_vars += extract_env_vars_inner(field_info.default_factory, name)
            else:
                if submodel_name:
                    field_name = submodel_name + env_settings.env_nested_delimiter + name
                else:
                    field_name = name
                env_var_name = env_settings._extract_field_info(field_info, field_name)[0][1]
                env_vars.append(env_var_name)
        return env_vars

    env_settings = EnvSettingsSource(
        model.__class__,
        case_sensitive=model.model_config.get("case_sensitive", None),
        env_prefix=model.model_config.get("env_prefix", None),
        env_nested_delimiter=model.model_config.get("env_nested_delimiter", None),
    )

    return extract_env_vars_inner(model, None)
        
```
